### PR TITLE
Added Proto -> OSS qsim converter.

### PR DESCRIPTION
**Pull request by MichaelBroughton**
Adds item 2 from #263.

Sorry for the size of this PR. It is very self contained and doesn't touch a lot of code.

Adds a parser for converting from our proto representation into the qsim format used here: https://github.com/quantumlib/qsim/blob/master/

### CODE REVIEW

File: tensorflow_quantum/core/src/circuit_parser_qsim.cc [hunk 1]
```diff
@@ -0,0 +1,373 @@
...
+  Status u;
+  unused = absl::SimpleAtoi(op.qubits(0).id(), &q0);
+  u = ParseProtoArg(op, "exponent", param_map, &exp);
+  if (!u.ok()) {
+    return u;
+  }
+  u = ParseProtoArg(op, "exponent_scalar", param_map, &exp_s);
+  if (!u.ok()) {
+    return u;
+  }
```
Comment (jaeyoo) [inline tensorflow_quantum/core/src/circuit_parser_qsim.cc:112]: I remember that there exists a gate with `exponent` having no `exponent_scalar` or vice versa in Cirq. These lines are guaranteed in qsim?

File: tensorflow_quantum/core/src/circuit_parser_qsim.cc [hunk 2]
```diff
@@ -0,0 +1,373 @@
...
+  return Status::OK();
+}
+
+// two qubit eigen -> Create(time, q0, q1, exp, gs)
+inline Status TwoEigenGate(
+    const Operation& op, const SymbolMap& param_map,
+    const std::function<QsimGate(unsigned, unsigned, unsigned, float, float)>&
+        create_f,
+    const unsigned time, QsimCircuit* circuit) {
+  int q0, q1;
```
Comment (jaeyoo) [inline tensorflow_quantum/core/src/circuit_parser_qsim.cc:128]: `create_f()` has unsigned qubit indices, but they are `int`. could you make them `unsigned int` or whatever `unsigned` type that you intended?

File: tensorflow_quantum/core/src/circuit_parser_qsim.cc [hunk 3]
```diff
@@ -0,0 +1,373 @@
...
+  circuit->gates.push_back(create_f(time, q0, q1));
+  return Status::OK();
+}
+
+// single qubit eigen -> Create(time, q0, exponent, global_shift)
+inline Status SingleEigenGate(
+    const Operation& op, const SymbolMap& param_map,
+    const std::function<QsimGate(unsigned, unsigned, float, float)>& create_f,
+    const unsigned time, QsimCircuit* circuit) {
+  int q0;
```
Comment (jaeyoo) [inline tensorflow_quantum/core/src/circuit_parser_qsim.cc:100]: `unsigned`?

File: tensorflow_quantum/core/src/circuit_parser_qsim.cc [hunk 4]
```diff
@@ -0,0 +1,373 @@
...
+  unused = absl::SimpleAtoi(op.qubits(1).id(), &q1);
+
+  u = ParseProtoArg(op, "exponent", param_map, &exp);
+  if (!u.ok()) {
+    return u;
+  }
+  u = ParseProtoArg(op, "exponent_scalar", param_map, &exp_s);
+  if (!u.ok()) {
+    return u;
+  }
```
Comment (jaeyoo) [inline tensorflow_quantum/core/src/circuit_parser_qsim.cc:142]: ditto.

File: tensorflow_quantum/core/src/circuit_parser_qsim.cc [hunk 5]
```diff
@@ -0,0 +1,373 @@
...
+  u = ParseProtoArg(op, "phase_exponent", param_map, &pexp);
+  if (!u.ok()) {
+    return u;
+  }
+  u = ParseProtoArg(op, "phase_exponent_scalar", param_map, &pexp_s);
+  if (!u.ok()) {
+    return u;
+  }
```
Comment (jaeyoo) [inline tensorflow_quantum/core/src/circuit_parser_qsim.cc:253]: ditto. `exponent` vs `exponent_scalar` and `phase_exponent` vs `phase_exponent_scalar`.

File: tensorflow_quantum/core/src/circuit_parser_qsim.cc [hunk 6]
```diff
@@ -0,0 +1,373 @@
...
+                      time, circuit);
+}
+
+Status ISwapGate(const Operation& op, const SymbolMap& param_map,
+                 const unsigned time, QsimCircuit* circuit) {
+  return TwoEigenGate(op, param_map, &qsim::Cirq::ISwapPowGate<float>::Create,
+                      time, circuit);
+}
+
+// single qubit PhasedXPow -> Create(time, q0, pexp, exp)
```
Comment (jaeyoo) [inline tensorflow_quantum/core/src/circuit_parser_qsim.cc:229]: `// single qubit PhasedXPow -> Create(time, q0, pexp, exp, gs)`

File: tensorflow_quantum/core/src/circuit_parser_qsim.h [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+#ifndef TFQ_CORE_SRC_CIRCUIT_PARSER_QSIM_H_
+#define TFQ_CORE_SRC_CIRCUIT_PARSER_QSIM_H_
+
+#include <string>
+#include <vector>
+
+#include "../qsim/lib/circuit.h"
+#include "../qsim/lib/fuser.h"
+#include "../qsim/lib/gates_cirq.h"
```
Comment (jaeyoo) [inline tensorflow_quantum/core/src/circuit_parser_qsim.h:24]: why `../qsim/*`, not `qsim/*`?

File: tensorflow_quantum/core/src/circuit_parser_qsim_test.cc [hunk 8]
```diff
@@ -0,0 +1,480 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "tensorflow_quantum/core/src/circuit_parser_qsim.h"
+
+#include <string>
+
+#include "../qsim/lib/circuit.h"
+#include "../qsim/lib/gates_cirq.h"
```
Comment (jaeyoo) [inline tensorflow_quantum/core/src/circuit_parser_qsim_test.cc:21]: `qsim/*`?

File: tensorflow_quantum/core/src/circuit_parser_qsim_test.cc [hunk 9]
```diff
@@ -0,0 +1,480 @@
...
+inline void AssertOneQubitEqual(const QsimGate& a, const QsimGate& b) {
+  for (int i = 0; i < 8; i++) {
+    ASSERT_NEAR(a.matrix[i], b.matrix[i], 1e-5);
+  }
+  ASSERT_EQ(a.qubits[0], b.qubits[0]);
+}
+
+class TwoQubitEigenFixture
+    : public ::testing::TestWithParam<std::tuple<
+          std::string, std::function<QsimGate(int, int, int, float, float)>>> {
```
Comment (jaeyoo) [inline tensorflow_quantum/core/src/circuit_parser_qsim_test.cc:72]: not `unsigned` here. we used `unsigned` in `circuit_parser_qsim.cc`.

File: tensorflow_quantum/core/src/circuit_parser_qsim_test.cc [hunk 10]
```diff
@@ -0,0 +1,480 @@
...
+      operations_proto->mutable_args();
+  (*args_proto)["global_shift"] = MakeArg(gs);
+  (*args_proto)["exponent"] = MakeArg("placeholder");
+  (*args_proto)["exponent_scalar"] = MakeArg(0.5);
+
+  // Set the qubits.
+  Qubit* qubits_proto = operations_proto->add_qubits();
+  qubits_proto->set_id("0");
+  qubits_proto = operations_proto->add_qubits();
+  qubits_proto->set_id("1");
```
Comment (jaeyoo) [inline tensorflow_quantum/core/src/circuit_parser_qsim_test.cc:103]: Could you add a comment describing the circuit diagram?

File: tensorflow_quantum/core/src/circuit_parser_qsim_test.cc [hunk 11]
```diff
@@ -0,0 +1,480 @@
...
+  ASSERT_EQ(
+      QsimCircuitFromProgram(program_proto, symbol_map, 2, &test_circuit,
+                             &fused_circuit),
+      tensorflow::Status(tensorflow::error::INVALID_ARGUMENT,
+                         "Could not find symbol in parameter map: alpha"));
+}
+
+TEST(QsimCircuitParserTest, PhasedISwapTest) {
+  float exponent = 0.1234;
+  float phase_exponent = 0.4567;
```
Comment (jaeyoo) [inline tensorflow_quantum/core/src/circuit_parser_qsim_test.cc:386]: PXP(PhasedXPowGate) is missed. I think the PXP(PhasedXPowGate) can be tested inside here with this PISP. Could you?

File: tensorflow_quantum/core/src/circuit_parser_qsim.cc [hunk 12]
```diff
@@ -0,0 +1,373 @@
...
+
+#include "tensorflow_quantum/core/src/circuit_parser_qsim.h"
+
+#include <string>
+#include <vector>
+
+#include "../qsim/lib/circuit.h"
+#include "../qsim/lib/fuser.h"
+#include "../qsim/lib/fuser_basic.h"
+#include "../qsim/lib/gates_cirq.h"
```
[the remark this answers is not in the document]
Comment (MichaelBroughton) [reply tensorflow_quantum/core/src/circuit_parser_qsim.cc:24]: I also found this weird. I think it's something to do with qsim not being setup just right ?

File: tensorflow_quantum/core/src/circuit_parser_qsim.cc [hunk 13]
```diff
@@ -0,0 +1,373 @@
...
+inline Status ParseProtoArg(const Operation& op, const std::string& arg_name,
+                            const SymbolMap& param_map, float* result) {
+  // find arg_name in proto.
+  const auto arg_v = op.args().find(arg_name);
+  if (arg_v == op.args().end()) {
+    return Status(tensorflow::error::INVALID_ARGUMENT,
+                  "Could not find arg: " + arg_name + " in op.");
+  }
+  // find proto arg field.
+  const auto proto_arg = arg_v->second;
```
[the remark this answers is not in the document]
Comment (MichaelBroughton) [reply tensorflow_quantum/core/src/circuit_parser_qsim.cc:54]: Done.

File: tensorflow_quantum/core/src/circuit_parser_qsim.cc [hunk 14]
```diff
@@ -0,0 +1,373 @@
...
+  *result = proto_arg.arg_value().float_value();
+  if (!proto_arg.symbol().empty()) {
+    // find symbol value in param_map.
+    const auto iter = param_map.find(proto_arg.symbol());
+    if (iter == param_map.end()) {
+      return Status(
+          tensorflow::error::INVALID_ARGUMENT,
+          "Could not find symbol in parameter map: " + proto_arg.symbol());
+    }
+    *result = iter->second.second;
```
[the remark this answers is not in the document]
Comment (MichaelBroughton) [reply tensorflow_quantum/core/src/circuit_parser_qsim.cc:64]: Done.

File: tensorflow_quantum/core/src/circuit_parser_qsim.cc [hunk 15]
```diff
@@ -0,0 +1,373 @@
...
+
+// series of fixed signature gate builders.
+// there is no need to error check for unparseable symbols
+// or proto args not being present. Those errors are caught
+// upstream.
+
+// single qubit gate Create(time, q0)
+inline Status SingleConstantGate(
+    const Operation& op, const SymbolMap& param_map,
+    const std::function<QsimGate(unsigned, unsigned)>& create_f,
```
[the remark this answers is not in the document]
Comment (MichaelBroughton) [reply tensorflow_quantum/core/src/circuit_parser_qsim.cc]: Fixed this and all unsigned to unsigned int below.

File: tensorflow_quantum/core/src/circuit_parser_qsim.cc (hunk 1 above)
Comment (MichaelBroughton) [reply tensorflow_quantum/core/src/circuit_parser_qsim.cc:116]: We actually add `exponent_scalar` in our `serializer.py` so it is gauranteed that we will have it.

File: tensorflow_quantum/core/src/circuit_parser_qsim.cc (hunk 2 above)
Comment (MichaelBroughton) [reply tensorflow_quantum/core/src/circuit_parser_qsim.cc:132]: Fixed.

File: tensorflow_quantum/core/src/circuit_parser_qsim.cc (hunk 3 above)
Comment (MichaelBroughton) [reply tensorflow_quantum/core/src/circuit_parser_qsim.cc]: Fixed.

File: tensorflow_quantum/core/src/circuit_parser_qsim.cc (hunk 6 above)
Comment (MichaelBroughton) [reply tensorflow_quantum/core/src/circuit_parser_qsim.cc]: Fixed.

File: tensorflow_quantum/core/src/circuit_parser_qsim.h (hunk 7 above)
Comment (MichaelBroughton) [reply tensorflow_quantum/core/src/circuit_parser_qsim.h:24]: qsim weirdly wouldn't let me do those imports. Wasn't sure how to fix right away. Do you have any ideas ?

File: tensorflow_quantum/core/src/circuit_parser_qsim_test.cc (hunk 11 above)
Comment (MichaelBroughton) [reply tensorflow_quantum/core/src/circuit_parser_qsim_test.cc:386]: Added.

File: tensorflow_quantum/core/src/circuit_parser_qsim_test.cc (hunk 10 above)
Comment (MichaelBroughton) [reply tensorflow_quantum/core/src/circuit_parser_qsim_test.cc:103]: What did you think we should add here could you help me understand ?

File: tensorflow_quantum/core/src/circuit_parser_qsim.cc (hunk 12 above)
[the remark this answers is not in the document]
Comment (jaeyoo) [reply tensorflow_quantum/core/src/circuit_parser_qsim.cc:24]: Let's skip this and make a new issue and PR for this.

File: tensorflow_quantum/core/src/circuit_parser_qsim.cc (hunk 1 above)
Comment (jaeyoo) [reply tensorflow_quantum/core/src/circuit_parser_qsim.cc:116]: Thanks a lot!

File: tensorflow_quantum/core/src/circuit_parser_qsim_test.cc (hunk 10 above)
Comment (jaeyoo) [reply tensorflow_quantum/core/src/circuit_parser_qsim_test.cc:103]: It was for easy maintenance of this test code. Cirq also drew circuit diagram for their test codes. But I feel it is time-consuming for this now... So let's ignore my thought :) closed.

File: tensorflow_quantum/core/src/circuit_parser_qsim.h (hunk 7 above)
Comment (jaeyoo) [reply tensorflow_quantum/core/src/circuit_parser_qsim.h:24]: Let's defer to #272 . Thanks!

**Pull request merged**